rust/agents/validator/src/submit.rs

```diff
@@ -110,40 +113,38 @@ impl ValidatorSubmitter {
                 .latest_checkpoint_observed
                 .set(latest_checkpoint.index as i64);
 
+            let tree_count = async {
+                let tree = tree.lock().await;
+                tree.count()
+            }
+            .await;
+
             if should_log_checkpoint_info() {
-                info!(
-                    ?latest_checkpoint,
-                    tree_count = tree.count(),
-                    "Latest checkpoint"
-                );
+                info!(?latest_checkpoint, tree_count, "Latest checkpoint");
             }
 
             // This may occur e.g. if RPC providers are unreliable and make calls against
             // inconsistent block tips.
             //
             // In this case, we just sleep a bit until we fetch a new latest checkpoint
             // that at least meets the tree.
-            if tree_exceeds_checkpoint(&latest_checkpoint, &tree) {
+            if tree_exceeds_checkpoint(&latest_checkpoint, tree_count) {
                 debug!(
                     ?latest_checkpoint,
-                    tree_count = tree.count(),
-                    "Latest checkpoint is behind tree, sleeping briefly"
+                    tree_count, "Latest checkpoint is behind tree, sleeping briefly"
                 );
                 sleep(self.interval).await;
                 continue;
             }
 
-            tree = call_and_retry_indefinitely(|| {
-                let mut tree = tree;
+            call_and_retry_indefinitely(|| {
                 let self_clone = self.clone();
+                let tree = tree.clone();
```

Comment:
I think this may introduce a new bug

imagine the tree has size 100 and the latest checkpoint is 110, but an error occurs part way through `submit_checkpoints_until_correctness_checkpoint`. I think there could be two types of errors: (1) the tree is only able to be built up to size 105, or (2) the whole tree is built up to size 110, but when the signed checkpoints are submitted to s3, only some are successfully submitted to s3 and then an error occurs (e.g. 110 - 106 (reverse order now) are successful, but 105 isn't).

We'd get an error from `submit_checkpoints_until_correctness_checkpoint` and then retry, but we'd retry building upon the tree that was modified in that last call to `submit_checkpoints_until_correctness_checkpoint`. So in error case 1 we'd start building from 105 to 110, and we'd skip signing checkpoints 101 - 105 entirely. And in error case 2 we'd start building the tree from 110, meaning we permanently skip the checkpoints 105 - 101 that we didn't get to submit last time.

One way around this might be to copy tree here, use the copied tree in `submit_checkpoints_until_correctness_checkpoint`, and then after we know that submitting all the checkpoints in that function is successful, we write / commit that new tree back to the Arc<Mutex>'d tree

I think this bug is also present for the backward one :/ not sure my suggestion above makes sense for the backward case, so something else may be needed

Reply:
fixed by retrying indefinitely at the individual checkpoint submission step, so we're fixing both the merkle tree being out of sync with what's submitted to checkpoint storage, and the submitter rewinds due to rate limits